python/private/attributes.bzl

```diff
@@ -406,35 +394,65 @@ as part of a runnable program (packaging rules may include them, however).
+            # Google builds change the set of allowed files.
+            allow_files = SRCS_ATTR_ALLOW_FILES,
+            # Necessary for --compile_one_dependency to work.
+            flags = ["DIRECT_COMPILE_TIME_INPUT"],
+            doc = """
+The list of Python source files that are processed to create the target. This
+includes all your checked-in code and may include generated source files.  The
+`.py` files belong in `srcs` and library targets belong in `deps`. Other binary
+files that may be needed at run time belong in `data`.
+""",
+        ),
         # NOTE: In Google, this attribute is deprecated, and can only
         # effectively be PY3 or PY3ONLY. Externally, with Bazel, this attribute
         # has a separate story.
-        # Required attribute, but the details vary by rule.
-        # Use create_srcs_version_attr to create one.
-        "srcs_version": None,
-        "_precompile_flag": attr.label(
+        ##"srcs_version": None,
+        "srcs_version": lambda: rule_builders.StringAttrBuilder(
+            doc = "Defunct, unused, does nothing.",
+        ),
+        "_precompile_flag": lambda: rule_builders.LabelAttrBuilder(
             default = "//python/config_settings:precompile",
             providers = [BuildSettingInfo],
         ),
-        "_precompile_source_retention_flag": attr.label(
+        "_precompile_source_retention_flag": lambda: rule_builders.LabelAttrBuilder(
             default = "//python/config_settings:precompile_source_retention",
             providers = [BuildSettingInfo],
         ),
         # Force enabling auto exec groups, see
         # https://bazel.build/extending/auto-exec-groups#how-enable-particular-rule
-        "_use_auto_exec_groups": attr.bool(default = True),
+        "_use_auto_exec_groups": lambda: rule_builders.BoolAttrBuilder(
+            default = True,
+        ),
     },
-    allow_none = True,
 )
 
+COVERAGE_ATTRS = {
+    # Magic attribute to help C++ coverage work. There's no
+    # docs about this; see TestActionBuilder.java
+    "_collect_cc_coverage": lambda: rule_builders.LabelAttrBuilder(
+        default = "@bazel_tools//tools/test:collect_cc_coverage",
+        executable = True,
+        cfg = "exec",
+    ),
+    # Magic attribute to make coverage work. There's no
+    # docs about this; see TestActionBuilder.java
+    "_lcov_merger": lambda: rule_builders.LabelAttrBuilder(
+        default = configuration_field(fragment = "coverage", name = "output_generator"),
+        executable = True,
+        cfg = "exec",
+    ),
+}
+
 # Attributes specific to Python executable-equivalent rules. Such rules may not
 # accept Python sources (e.g. some packaged-version of a py_test/py_binary), but
 # still accept Python source-agnostic settings.
-AGNOSTIC_EXECUTABLE_ATTRS = union_attrs(
+AGNOSTIC_EXECUTABLE_ATTRS = dicts.add(
     DATA_ATTRS,
     {
-        "env": attr.string_dict(
+        "env": lambda: rule_builders.StringDictAttrBuilder(
```

Comment:
Nit, instead of rule_builders, should we call it pyrule_builders? Is there anything python specific in our code? If not, we could keep it outside //python, similarly to how sphinxdocs live.

Reply:
I am not sure about the maintenance status of `bazel_skylib`, but it does seem that having some of these copied/published when they are mature could be handy. 

That said, we can do the refactor later and move `repo_utils` and `text_utils` and this into this new location.